### PR TITLE
Features/player2

### PR DESCRIPTION
**Pull request by ghost**
https://trello.com/c/1GzCiTbh/30-1-as-norbert-i-want-to-play-as-player-2

image

- [x] integrate a second player
- [x] make a loop for the prompt to handle players turn 
- [x] update display to show player information
- [x] doc
- [x] tests

### CODE REVIEW

File: src/prompt.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  output: process.stdout,
+});
+
+function question(text: string) {
+  return new Promise((resolve) => {
+    rl.question(text, resolve);
+  });
+}
+
+export async function readNextPlay(gameState: GameState): Promise<number> {
```
Comment (atilbian) [inline src/prompt.ts]:
```suggestion
export async function readNextPlay(): Promise<number> {
```

Useless param here

File: src/prompt.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import * as readline from "readline";
+import { messages, PLACEHOLDER } from "./config/messages.js";
+import { getPlayerTokenChar, boardLayout } from "./layout/cliLayout.js";
+import { GameState, PlayerNum } from "./types/gameState.js";
+const rl: readline.Interface = readline.createInterface({
```
Comment (atilbian) [inline src/prompt.ts:6]:
```suggestion

const rl: readline.Interface = readline.createInterface({
```

Sorry for my psychorigidity again

File: src/tools.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export function transpose(matrix: Array<Array<number>>) {
+  return matrix[0].map((_, colIndex) => matrix.map((row) => row[colIndex]));
+}
```
Comment (atilbian) [inline src/tools.ts:3]: **suggestion(non-blocking):** I didn't see that there were no unit tests for this function, it would be nice to add some, but it's totally not a priority.

File: src/layout/cliLayout.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  gameState.boardState.forEach((line: Array<number>) => {
+    line.forEach((token: number) => {
+      resultDisplay.push(
+        `${boardLayout.SEPARATOR} ${
+          token === 1
+            ? boardLayout.PLAYER_ONE_TOKEN
+            : token === 2
+              ? boardLayout.PLAYER_TWO_TOKEN
+              : boardLayout.EMPTY_TOKEN
```
Comment (atilbian) [inline src/layout/cliLayout.ts]: **suggestion:** You can use getPlayerTokenChar instead here nop ?

File: src/connect4.ts
Comment (atilbian) [inline src/connect4.ts:1]: I didn't see the return of `countNbTokens` and I'm not a big fan of that, I prefer to return an object instead so we can be sure which value belongs to what:
for example {p1Count:11, p2Count:11, emptyCount: 20} instead of [20,11,11]

It could help us avoid errors later

File: src/connect4.ts [hunk 5]
```diff
@@ -1,95 +1,24 @@
...
  */
-export function playToken(
-  boardState: BoardState,
-  column: number,
-  pNum: number,
-): BoardState {
-  if (column < 0 || column >= boardLayout.NB_COLUMN) {
-    throw new Error(boardLayout.ERROR_INVALID_COLUMN_NUMBER);
+export function playToken(gameState: GameState, column: number): GameState {
+  if (column < 0 || column > boardLayout.NB_COLUMN) {
```
Comment (atilbian) [inline src/connect4.ts]:
```suggestion
  if (column <= 0 || column > boardLayout.NB_COLUMN) {
```
As the first column = 1 and not 0

**Review by atilbian — CHANGES_REQUESTED** [review]
**praise:** Well done, it's good to see that you've started splitting the big file, it will help for later reviews and developments

File: src/prompt.ts (hunk 2 above)
Comment (ghost) [reply src/prompt.ts:6]: I'm too ... I understand you ...

File: src/layout/cliLayout.ts (hunk 4 above)
Comment (ghost) [reply src/layout/cliLayout.ts]: trully

File: src/connect4.ts (hunk 5 above)
Comment (ghost) [reply src/connect4.ts]: rho

File: src/connect4.spec.ts [hunk 6]
```diff
@@ -26,7 +26,11 @@ test("countNbToken", () => {
     [0, 2, 2, 1, 1, 2, 2],
   ];
 
-  expect(countNbTokens(boardState)).toEqual([20, 11, 11]);
+  expect(countNbTokens(boardState)).toEqual({
+    emptyCount: 20,
+    p1Count: 11,
+    p2Count: 11,
+  });
```
Comment (atilbian) [inline src/connect4.spec.ts:33]: Noiiiiiiiiiiice

File: src/tools.spec.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    [0, 0, 2, 2, 1, 2],
+    [0, 0, 1, 1, 1, 2],
+    [0, 0, 0, 1, 2, 1],
+    [1, 2, 2, 2, 1, 1],
+    [0, 0, 0, 0, 1, 2],
+    [0, 0, 0, 2, 1, 2],
+  ];
+
+  expect(transpose(boardState)).toEqual(tBoardState);
+});
```
Comment (atilbian) [inline src/tools.spec.ts:24]: Nice ! Some tests could be added (for edge cases like a lines of differents sizes for example) but this is for later.

**Review by atilbian — APPROVED** [review]
